### PR TITLE
Start/stop components in a more synchronised manner

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

Two things being fixed here:

First one is order of start/stop. 
Currently we start new components before stopping old. This is usually fine unless component is a service (start/stop is slower).
When new service is started old one can be still running and using resources needed for new service to start. 
We can than see something like `bind port already in use`.

Second thing fixed here is waiting for uninstall to finish.
Again problem is when we change output and we try to uninstall component `service-output-one` and start `service-output-two`
We kick off uninstall of `service-output-one` and we actually don't wait for operation to finish. Then we start `service-output-two`
This may be fine as `service-output-one` may already be stopped but uninstall may be still in progress and this uninstall will remove service even though it's needed for `service-output-two`  

## Why is it important?

Service failures when
- we swap outputs
- we upgrade agent.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

### CODE REVIEW

File: pkg/component/runtime/manager.go [hunk 1]
```diff
@@ -646,40 +649,81 @@ func (m *Manager) update(components []component.Component, teardown bool) error
...
 		// skip if already touched (meaning it still existing)
 		if _, done := touched[id]; done {
 			continue
 		}
 		// component was removed (time to clean it up)
 		_ = existing.stop(teardown)
+		// stop is async, wait for operation to finish,
+		// otherwise new instance may be started and components
+		// may fight for resources (e.g ports, files, locks)
+		m.waitForStopped(existing)
```
Comment (pchila) [inline pkg/component/runtime/manager.go:676]: Nitpicking here: can't we stop and wait for stopped services in parallel using goroutines and a WaitGroup? This would allow us to be a bit faster instead of shutting down one component at a time... Maybe this can be a further optimization down the line ?

**Review by pchila — APPROVED** [review]

File: pkg/component/runtime/manager.go (hunk 1 above)
Comment (blakerouse) [reply pkg/component/runtime/manager.go:676]: Yes this should absolutely happen in parallel.

**Review by blakerouse — CHANGES_REQUESTED** [review]
This should handle the stopped in parallel, as @pchila stated.

This also really needs to have a test. We have many tests that cover the runtime manager and this being critical behavior this needs to be tested and covered.

Should be easy to subscribe to component state changes from the runtime and watch that the all components that should be stopped are stopped before even spawning the new components.

**Comment by michalpristas** [comment]
@blakerouse added parallelism and test

**Review by blakerouse — APPROVED** [review]
Thanks for adding the parallelism and the test!

**Comment by jlind23** [comment]
@pchila @blakerouse where you able to test this new behaviour locally? @cmacknz @michalpristas any corner cases we might have missed?
